### PR TITLE
Small improvements to authentication

### PR DESCRIPTION
**Pull request by kevinrobinson**
This is intended to be merged into https://github.com/mit-teaching-systems-lab/swipe-right-for-cs/pull/93, and makes a few small improvements.

### 1. Remove linkToken from browser URL bar
This reads the token in right away, and then uses `replaceState` to remove it from the user's browser bar, making the token less visible.

### 2. In server responses, don't reveal whether an email is in the whitelist
This updates the UI message as well:
screen shot 2018-01-29 at 11 14 50 am

### 3. Add rate limiter for sensitive endpoints
To prevent naive brute forcing.

### 4. Add server logging about uauthorized attempts and access
eg: 
```
Unauthorized access attempt, email: fds@mit.edu, linkToken: e6c07719-6b30-4375-ae66-7222112f514b
Unauthorized access attempted by email:  fdsoo@mit.edu
Accessed with linkToken:  49213d9e-749b-4d5e-b520-1762624e3f8a
Granted access token...
```

### CODE REVIEW

**Comment by kevinrobinson** [comment]
@keving17 Can you review these and see if they seem reasonable?  And then I'll merge this in and deploy later today.

File: client/src/research/EmailLinkLoginPage.js [hunk 1]
```diff
@@ -12,32 +13,36 @@ class EmailLinkLoginPage extends Component {
...
 
     this.onUpdateEmail = this.onUpdateEmail.bind(this);
     this.onSubmit = this.onSubmit.bind(this);
-    this.getQueryVariable = this.getQueryVariable.bind(this);
     this.authenticate = this.authenticate.bind(this);
   }
 
-  getQueryVariable(variable) {
+  // Also remove login token from URL bar in browser
+  readLinkToken() {
```
Comment (keving17) [inline client/src/research/EmailLinkLoginPage.js]: looks good! 👍

File: client/src/research/EmailLinkLoginPage.js [hunk 2]
```diff
@@ -51,24 +56,19 @@ class EmailLinkLoginPage extends Component {
...
-    this.authenticate(linkToken)
-      .then(result => {
-        this.setState({token:result.token});
-        this.onSubmitSuccess();
-      })
-      .catch(err => {
-        this.onSubmitError();
-      });
+    this.authenticate()
+      .then(result => this.onSubmitSuccess(result.token))
```
Comment (keving17) [inline client/src/research/EmailLinkLoginPage.js]: Nice simplification here!

File: client/src/research/LoginPage.js [hunk 3]
```diff
@@ -34,14 +34,11 @@ class LoginPage extends Component {
       })
     })
       .then(result => {
-        if (result.status === 200) {
-          this.setState({message: "Check your email, "+this.state.email+" for a login link!"});
-        } else{
-          this.setState({message: "Your email, "+this.state.email+", does not seem to be authorized to view user data."});
-        }
+        // The server shouldn't leak whether it's a whitelisted email address or not
+        this.setState({ message: "Check your email.  If you are authorized, you should see a login link within a few seconds!"});
```
Comment (keving17) [inline client/src/research/LoginPage.js]: This makes sense. I hadn't thought of this security flaw. 💯

**Comment by kevinrobinson** [comment]
@keving17 Great, thanks for reviewing!  I pushed one more commit that adds a global kill switch for researcher access, and also limits the email link time window to 5min.

**Pull request merged**